### PR TITLE
[FIXED JENKINS-21668] Obtain root CA information for java instances

### PR DESCRIPTION
**Pull request by christ66**
This will only work if there is no password for obtaining the keystore.

### CODE REVIEW

File: src/main/java/com/cloudbees/jenkins/support/impl/RootCAs.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    } else {
+      name = "slave/" + node.getNodeName();
+    }
+    container.add(
+            new Content("nodes/" + name + "/RootCA.txt") {
+              @Override
+              public void writeTo(OutputStream os) throws IOException {
+                PrintWriter out = new PrintWriter(new BufferedWriter(new OutputStreamWriter(os, "utf-8")));
+                try {
+                  out.println(getRootCA(node));
```
Comment (christ66) [inline src/main/java/com/cloudbees/jenkins/support/impl/RootCAs.java]: I removed the call to `Future<String>`.

File: src/main/java/com/cloudbees/jenkins/support/impl/RootCAs.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+      instance = KeyStore.getInstance(KeyStore.getDefaultType());
+      while (instance.aliases().hasMoreElements()) {
+        String s = instance.aliases().nextElement();
+        writer.append("========");
+        writer.append("Alias: " + s);
+        writer.append(instance.getCertificate(s).getPublicKey().toString());
+        writer.append("Trusted certificate: " + instance.isCertificateEntry(s));
+      }
+    } catch (KeyStoreException e) {
+      writer.write(ExceptionUtils.getFullStackTrace(e));
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/impl/RootCAs.java]: There is a similar method in `hudson.Functions` BTW.

File: src/main/java/com/cloudbees/jenkins/support/impl/RootCAs.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+    private static final long serialVersionUID = 1L;
+  }
+
+  public static void getRootCAList(StringWriter writer) {
+    KeyStore instance = null;
+    try {
+      instance = KeyStore.getInstance(KeyStore.getDefaultType());
+      while (instance.aliases().hasMoreElements()) {
+        String s = instance.aliases().nextElement();
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/impl/RootCAs.java]: This is wrong. Each call to `aliases()` creates a new `Enumeration`. So I would expect this to be an infinite loop, if there is at least one alias.

File: src/main/java/com/cloudbees/jenkins/support/impl/RootCAs.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        writer.append("========");
+        writer.append("Alias: " + s);
+        writer.append(instance.getCertificate(s).getPublicKey().toString());
+        writer.append("Trusted certificate: " + instance.isCertificateEntry(s));
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/impl/RootCAs.java]: Why are you repeating these last two things for each alias? Should be outside the loop I think.

Comment (christ66) [inline src/main/java/com/cloudbees/jenkins/support/impl/RootCAs.java]: Not sure I understand. Obtaining the certificate requires the alias, so the last two things should be in the loop.

**Comment by christ66** [comment]
@reviewbybees

**Comment by ndeloof** [comment]
:bee: :+1:

**Comment by stephenc** [comment]
:bee:

**Comment by stephenc** [comment]
@reviewbybees done (as @christ66 is on vacation)

**Comment by stephenc** [comment]
(wearing plugin maintainer's hat) :+1:

**Pull request merged**